### PR TITLE
Add timeout to npm install

### PR DESCRIPTION
**Pull request by johnpmitsch**
We are seeing npm install timeout in multiple katello dev environment runs

### CODE REVIEW

File: manifests/setup.pp [hunk 1]
```diff
@@ -25,7 +25,9 @@
   katello_devel::bundle { 'install --without mysql:mysql2 --retry 3 --jobs 3 --path .vendor':
     environment => ['MAKEOPTS=-j'],
   } ->
-  katello_devel::bundle { 'exec npm install': } ->
+  katello_devel::bundle { 'exec npm install':
+    timeout => '600'
```
[the remark this answers is not in the document]
Comment (ekohl) [reply manifests/setup.pp]: Looks like I had a draft review that I forgot to submit

[the remark this answers is not in the document]
Comment (johnpmitsch) [reply manifests/setup.pp]: I'm open to it, but if it is just the devel installer, I wonder how configurable we want it to be? I would think we just want to give npm install enough time to complete and devs shouldn't really care about a timeout but maybe I'm missing something

[the remark this answers is not in the document]
Comment (ekohl) [reply manifests/setup.pp]: My experience is that you almost always run into the point where you want to change it. For example, you need it to be 700. Making it a class parameter allows you to easily tune it without changing the source. So more of a good practice that I've learned to use the hard way.

**Comment by johnpmitsch** [comment]
@ekohl I updated to use a parameter, I'm not too familiar with puppet so hopefully this is what you meant :) I also updated the default to be 900, which is 15 minutes

File: manifests/init.pp [hunk 2]
```diff
@@ -100,12 +100,15 @@
   Integer[0, 1000] $qpid_wcache_page_size = $katello_devel::params::qpid_wcache_page_size,
   Array[String] $extra_plugins = $katello_devel::params::extra_plugins,
   String $rails_command = $katello_devel::params::rails_command,
+  Integer[0, 3600] $npm_timeout = $katello_devel::params::npm_timeout,
```
Comment (ekohl) [inline manifests/init.pp]: You don't need to set an upper limit

```suggestion
  Integer[0] $npm_timeout = $katello_devel::params::npm_timeout,
```

File: manifests/init.pp [hunk 3]
```diff
@@ -100,12 +100,15 @@
...
 ) inherits katello_devel::params {
 
   $fork_remote_name_real = pick_default($fork_remote_name, $github_username)
 
   $foreman_dir = "${deployment_dir}/foreman"
 
+  $npm_timeout = 900
```
Comment (ekohl) [inline manifests/init.pp]: This should be in params.pp

**Comment by johnpmitsch** [comment]
@ekohl thanks, updated!

File: manifests/setup.pp [hunk 4]
```diff
@@ -25,7 +26,9 @@
   katello_devel::bundle { 'install --without mysql:mysql2 --retry 3 --jobs 3 --path .vendor':
     environment => ['MAKEOPTS=-j'],
   } ->
-  katello_devel::bundle { 'exec npm install': } ->
+  katello_devel::bundle { 'exec npm install':
+    timeout => $npm_timeout,
```
Comment (ekohl) [inline manifests/setup.pp]: Test failure is relevant:

```
       Evaluation Error: Error while evaluating a Resource Statement, Katello_devel::Bundle[exec npm install]: has no parameter named 'timeout' (file: /home/travis/build/theforeman/puppet-katello_devel/spec/fixtures/modules/katello_devel/manifests/setup.pp, line: 29) on node travis-job-d9b6a10e-5bb7-499e-a206-ee586ce9ecf7.c.travis-ci-prod-2.internal
```

Looks like you'll need to add the parameter to https://github.com/theforeman/puppet-katello_devel/blob/master/manifests/bundle.pp

**Comment by johnpmitsch** [comment]
@ekohl this is getting pretty complicated for one parameter

> It could have been sufficient to only add it to setup.pp 

Is this a better approach?

**Comment by ekohl** [comment]
> @ekohl this is getting pretty complicated for one parameter
> 
> > It could have been sufficient to only add it to setup.pp
> 
> Is this a better approach?

You already have the code now, but it would have been easier to do with in the end sufficient flexibility. You wouldn't have needed to add it to `init.pp` and `params.pp`.

File: manifests/bundle.pp [hunk 5]
```diff
@@ -8,6 +8,7 @@
   $use_scl_ruby = $katello_devel::use_scl_ruby,
   $user = $katello_devel::user,
   $cwd = $katello_devel::foreman_dir,
+  $npm_timeout = $katello_devel::npm_timeout,
```
Comment (ekohl) [inline manifests/bundle.pp]:
```suggestion
  Integer[0] $timeout = 600,
```

You'll also need to change line 33 and and use it there

File: manifests/bundle.pp [hunk 6]
```diff
@@ -29,8 +30,9 @@
     cwd         => $cwd,
     user        => $user,
     logoutput   => 'on_failure',
-    timeout     => '600',
+    timeout     => $timeout,
     path        => $path,
     unless      => $unless,
+    npm_timeout => $npm_timeout
```
Comment (ekohl) [inline manifests/bundle.pp]: This does not exist: https://puppet.com/docs/puppet/latest/type.html#exec-attributes
```suggestion
```

**Comment by johnpmitsch** [comment]
Looks like the failures are all the `::ipaddress` tests now

**Comment by ekohl** [comment]
https://github.com/theforeman/puppet-foreman_proxy/pull/537 should fix that.

**Comment by ekohl** [comment]
Restart the build now that that's merged.

**Pull request merged**